packages/loader/container-loader/src/deltaManager.ts

```diff
@@ -99,6 +102,38 @@ export interface IDeltaManagerInternalEvents extends IDeltaManagerEvents {
     (event: "closed", listener: (error?: ICriticalContainerError) => void);
 }
 
+/**
+ * Implementation of IDocumentDeltaConnection that does not support submitting
+ * or receiving ops. Used in storage-only mode.
+ */
+class NoDeltaStream extends TypedEventEmitter<IDocumentDeltaConnectionEvents> implements IDocumentDeltaConnection {
+    clientId: string = undefined as any;
+    claims: ITokenClaims = {
+        scopes: [ScopeType.DocRead],
+    } as any;
+    mode: ConnectionMode = "read";
+    existing: boolean = true;
+    maxMessageSize: number = 0;
+    version: string = "";
+    initialMessages: ISequencedDocumentMessage[] = [];
+    initialSignals: ISignalMessage[] = [];
+    initialClients: ISignalClient[] = [];
+    serviceConfiguration: IClientConfiguration = undefined as any;
+    checkpointSequenceNumber?: number | undefined = undefined;
+    submit(messages: IDocumentMessage[]): void {
+        const err = new Error("Cannot submit with storage only connection");
```

Comment:
I'd rather keep it conforming to existing processes. Existing workflow - submit() is no-op, but server could nack connection with appropriate error and message. That's how connection to read-only file works. See getNackReconnectInfo() and nack handling in DeltaManager, as well as reconnectOnError() checking for canRetryOnError(error)

Reply:
are you suggesting this should emit a nack on submit? rather than emit its own error?

Reply:
Correct, it's better for it to follow existing paradigms, and raise nack instead.
Code here is fine, it's just that I'd prefer to keep fewer paths that can generate errors - today submit is fire-and-forget (other than bugs in driver / delta manager).

Reply:
That said, I think the current design here is that we should never get here (as container runtime knows about this special state), vs. I advocate for container runtime does not need to know about this state (and thus we can get here).

If that's the case (we should never get here), then current design is fine. I'd still advocate to not add new properties and just expose different read-only modes through single property - container runtime could differentiate based on that.
My only reservation - it assumes that container runtime is the only one who can send ops. It's sort of true today (noop sending in loader layer exists, but it's only engaged in "write" mode).